File: internship_backend_apr25/jury_solutions/go/d.go
package main

import (
	"fmt"
	"math"
	"bufio"
	"os"
)

func main() {
	solve()
}

func solve() {
    reader := bufio.NewReader(os.Stdin)
    
	var n, m, k int
	fmt.Fscan(reader, &n, &m, &k)
	k++
	graph := make([][]struct {
		v int
		c int
	}, n)
	mx := 0

	for i := 0; i < m; i++ {
		var u, v, c int
		fmt.Fscan(reader, &u, &v, &c)
		u--
		v--
		graph[u] = append(graph[u], struct {
			v int
			c int
		}{v, c})
		mx = int(math.Max(float64(mx), float64(c)))
	}

	l, r := 1, mx+2
	ans := -1

	for l <= r {
		mid := (l + r) / 2

		if check(graph, k, mid) {
			r = mid - 1
			ans = mid
		} else {
			l = mid + 1
		}
	}

	fmt.Println(ans)
}

func checkGraph(graph [][]struct {
	v int
	c int
}, border int, order *[]int) bool {
	colors := make([]int, len(graph))
	res := false
	color := 1

	for i := range graph {
		if colors[i] != 0 {
			continue
		}
		res = res || topSortNoRec(graph, i, border, colors, order, color)
		color++
	}

	return res
}

func topSortNoRec(graph [][]struct {
	v int
	c int
}, v, border int, colors []int, order *[]int, curColor int) bool {
	dfs := []struct {
		visited bool
		node    int
	}{{false, v}}
	visited := make(map[int]struct{})
	visited[v] = struct{}{}

	for len(dfs) > 0 {
		node := dfs[len(dfs)-1]
		dfs = dfs[:len(dfs)-1]

		if node.visited {
			*order = append(*order, node.node)
			colors[node.node] = math.MinInt32
			continue
		}

		if colors[node.node] != 0 {
			continue
		}

		colors[node.node] = curColor
		dfs = append(dfs, struct {
			visited bool
			node    int
		}{true, node.node})

		for _, u := range graph[node.node] {
			if u.c > border {
				continue
			}

			if colors[u.v] == curColor {
				return true
			}

			if colors[u.v] == 0 {
				if _, ok := visited[u.v]; !ok {
					dfs = append(dfs, struct {
						visited bool
						node    int
					}{false, u.v})
					visited[u.v] = struct{}{}
				}
			}
		}
	}

	return false
}

func checkKLenPath(graph [][]struct {
	v int
	c int
}, k, border int, order []int) bool {
	if len(order) == 0 {
		return false
	}
	reverse(order)

	dp := make([]int, len(graph))
	maxLen := 0

	for _, v := range order {
		if dp[v] == 0 {
			dp[v] = 1
			maxLen = int(math.Max(float64(maxLen), 1))
		}

		for _, u := range graph[v] {
			if u.c > border {
				continue
			}
			dp[u.v] = int(math.Max(float64(dp[u.v]), float64(dp[v]+1)))
			maxLen = int(math.Max(float64(maxLen), float64(dp[u.v])))
		}
	}

	return maxLen >= k
}

func check(graph [][]struct {
	v int
	c int
}, k, border int) bool {
	var order []int
	return checkGraph(graph, border, &order) || checkKLenPath(graph, k, border, order)
}

func reverse(s []int) {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		s[i], s[j] = s[j], s[i]
	}
}

File: internship_backend_apr25/jury_solutions/cpp/d.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;

mt19937 rnd(chrono::steady_clock::now().time_since_epoch().count());

bool check_cycle (vector<vector<pair<int, ll>>>& graph, int v, ll border, vector<int>& colors)
{
	bool res = false;
	colors[v] = 1;
	for (auto& u : graph[v])
	{
		if (u.second > border) 
			continue;
		switch (colors[u.first])
		{
			case 1:
				return true;
			case 2:
				continue;
			default:
				res = res || check_cycle(graph, u.first, border, colors);
				break;
		}
	}
	colors[v] = 2;
	return res;
}

bool has_cycle (vector<vector<pair<int, ll>>>& graph, ll border)
{
	vector<int> colors(graph.size(), 0);

	bool res = false;

	for (int i = 0; i < graph.size(); i++)
	{
		if (colors[i] != 0) continue;

		res = res || check_cycle(graph, i, border, colors);
	}

	return res;
}

void top_sort (vector<vector<pair<int, ll>>>& graph, int v, ll border, vector<int>& order, vector<char>& visited)
{
	for (auto& u : graph[v])
	{
		if (visited[u.first] || u.second > border) 
		continue;
		visited[u.first] = true;
		top_sort(graph, u.first, border, order, visited);
	}

	order.emplace_back(v);
}

bool check_k_len_path (vector<vector<pair<int, ll>>>& graph, ll k, ll border)
{
	vector<int> order;
	vector<char> visited(graph.size(), 0);

	for (int i = 0; i < graph.size(); i++)
	{
		if (visited[i])
			continue;

		visited[i] = true;
		top_sort(graph, i, border, order, visited);
	}

	if (order.empty()) 
		return false;

	reverse(order.begin(), order.end());

	vector<ll> dp(graph.size(), 0);

	for (auto& v : order)
	{
		if (dp[v] == 0)
			dp[v] = 1;
		for (auto &u: graph[v])
		{
			if (u.second > border)
				continue;
			dp[u.first] = max(dp[u.first], dp[v] + 1);
		}
	}

	ll mx = 0;

	for (auto& i : dp)
		mx = max(mx, i);

	return mx >= k;
}

bool check (vector<vector<pair<int, ll>>>& graph, ll k, ll border)
{
	return has_cycle(graph, border) || check_k_len_path(graph, k, border);
}

int main () 
{
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);
	cin.sync_with_stdio(0);
	cout.sync_with_stdio(0);
	srand(time(0));
	cout.precision(20);

	ll n,m,k;
	cin >> n >> m >> k;
	k++;

	ll mx = 0;

	vector<vector<pair<int, ll>>> graph(n, vector<pair<int, ll>>());

	for (int i = 0; i < m; i++)
	{
		int u, v;
		ll cost;
		cin >> u >>v >> cost;
		mx = max(mx ,cost);
		u--;
		v--;
		graph[u].emplace_back(v, cost);
	}

	ll l = -1, r = mx + 10;
	ll ans = -1;

	while (l < r - 1)
	{
		ll mid = (l + r) / 2;

		if (check(graph, k, mid))
		{
			r = mid;
			ans = mid;
		}
		else
			l = mid;
	}

	cout << ans;

	return 0;
}
